app/controllers/bots_controller.rb

```diff
@@ -122,6 +126,45 @@ def remove_collaborator
   end
 
 
+
+  def get_data
+    render body: @bot_data.data, content_type: "application/octet-stream"
+  end
+
+  def update_data
+    if @bot_data.nil?  # insert a new BotData
+      @bot_data = BotData.new(bot: @bot, key: params[:data_key], data: request.raw_post)
+      @bot_data.save
+    else
+      @bot_data.update(data: request.raw_post)
+    end
+
+    head :no_content
+  end
+
+  def remove_data
+    @bot_data.destroy
+    head :no_content
+  end
+
+  def web_remove_data
+    unless current_user.is_owner?(@bot) || current_user.is_collaborator?(@bot) ||current_user.is_admin?
+      respond_to do |format|
+        format.html { redirect_to bots_path(@bot), flash: { error: 'You do not have permission to remove bot data.' } }
+      end
+    end
+
+     BotData.where(bot: @bot, key: params[:data_key]).first!.destroy
+
+     respond_to do |format|
+       format.html { redirect_to edit_bot_path(@bot), flash: { success: 'Data was successfully removed.' } }
```

Comment:
You don't actually know whether the data was removed successfully or not at this point - check the return value of `.destroy` to check that.